localstack/services/generic_proxy.py

```diff
@@ -672,56 +674,88 @@ def build_forward_url(self, host, path):
         return urlparse(forward_url)
 
 
-class EndpointProxy(ProxyListener):
-    def __init__(self, base_url: str, forward_url: str) -> None:
-        super().__init__()
-        self.forwarder = UrlMatchingForwarder(
-            base_url=base_url,
-            forward_url=forward_url,
-        )
+class RegisteredProxyServer(Server, abc.ABC):
+    """
+    A proxy server that can register itself at the edge router
+    """
 
-    def forward_request(self, method, path, data, headers):
-        return self.forwarder.forward_request(method, path, data, headers)
+    def __init__(self, base_url):
+        self._url = urlparse(base_url)
+        super().__init__(self._url.port, f"{self._url.hostname}")
+        self.rules = []
+
+    def register(self, forward_url):
+        """
+        Registers two routes at the edge router to implement either a path or a domain based routing.
+        We need two rules to match 1) the root path + no additional path 2) the root path + any other path.
+        This function is NOT called automatically, it is usually called in the leaf class' start up routine.
+
+        :param forward_url: The url which the proxy forwards to (from the initially passed base_url)
+        """
+
+        #   We MUST NOT create a catch all traffic rule
+        assert not (
+            (self._url.path == "" or self._url.path is None or self._url.path == "/")
+            and self._url.hostname == config.LOCALSTACK_HOSTNAME
+        )
```

Comment:
Is an 'assert' the proper way to go here? In terms of error handling, this will basically just "blow up", do we need some "more graceful" way? But I am not quite sure how we would handle this situation, because once we were to create such a catch all rule, LocalStack would become more or less unusable anyway.

Reply:
This is just a command which ensures that a runtime state which should never happen does never happen. So, in my opinion it's okay to have the assert here just to make sure this does not happen (and simplify the debugging in case it does happen).